services/s3/src/main/java/software/amazon/awssdk/services/s3/parsing/S3Uri.java

```diff
@@ -0,0 +1,253 @@
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * Object that represents a parsed S3 URI. Can be used to easily retrieve the bucket, key, region, style, and query parameters
+ * of the URI. Only path-style and virtual-hosted-style URI parsing is supported, including CLI-style URIs, e.g.,
+ * "s3://bucket/key". AccessPoints and Outposts URI parsing is not supported. If you work with object keys and/or query
+ * parameters with special characters, they must be URL-encoded, e.g., replace " " with "%20". If you work with
+ * virtual-hosted-style URIs with bucket names that contain a dot, i.e., ".", the dot must not be URL-encoded. Encoded buckets,
+ * keys, and query parameters will be returned decoded.
+ */
+@Immutable
+@SdkPublicApi
+public final class S3Uri implements ToCopyableBuilder<S3Uri.Builder, S3Uri> {
+
+    private final URI uri;
+    private final String bucket;
+    private final String key;
+    private final Region region;
+    private final boolean isPathStyle;
+    private final Map<String, List<String>> queryParams;
+
+    private S3Uri(Builder builder) {
+        this.uri = Validate.notNull(builder.uri, "URI must not be null");
+        this.bucket = builder.bucket;
+        this.key = builder.key;
+        this.region = builder.region;
+        this.isPathStyle = Validate.notNull(builder.isPathStyle, "Path style flag must not be null");
+        this.queryParams = Validate.notNull(builder.queryParams, "Query parameters map must not be null");
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    @Override
+    public Builder toBuilder() {
+        return new Builder(this);
+    }
+
+    /**
+     * Returns the original URI that was used to instantiate the {@link S3Uri}
+     */
+    public URI uri() {
+        return uri;
+    }
+
+    /**
+     * Returns the bucket specified in the URI. Returns an empty optional if no bucket is specified.
+     */
+    public Optional<String> bucket() {
+        return Optional.ofNullable(bucket);
+    }
+
+    /**
+     * Returns the key specified in the URI. Returns an empty optional if no key is specified.
+     */
+    public Optional<String> key() {
+        return Optional.ofNullable(key);
+    }
+
+    /**
+     * Returns the region specified in the URI. Returns an empty optional if no region is specified, i.e., global endpoint.
+     */
+    public Optional<Region> region() {
+        return Optional.ofNullable(region);
+    }
+
+    /**
+     * Returns true if the URI is path-style, false if the URI is virtual-hosted style.
+     */
+    public boolean isPathStyle() {
+        return isPathStyle;
+    }
+
+    /**
+     * Returns a map of the query parameters specified in the URI. Returns an empty map if no queries are specified.
+     */
+    public Map<String, List<String>> rawQueryParameters() {
+        return Collections.unmodifiableMap(queryParams);
```

Comment:
We probably should do a deep copy. https://github.com/aws/aws-sdk-java-v2/blob/master/utils/src/main/java/software/amazon/awssdk/utils/CollectionUtils.java#L83

For performance optimization, can we deep copy it in the ctor instead so that we just do it once?

Reply:
+1